### DISCUSSION TITLE
Ports are not assigned correctly

### DISCUSSION THREAD

**Discussion by deathwaiting** [acorn-io/runtime#745] [General]
I am kubernetes noob, and currently trying to run a `minikube` cluster on local machine as dependencies for my app, using.
I am trying to publish ports 5432 and 8080, hopefully on localhost, or on the cluster IP, but the services are assigned to random ports instead

my acorn file

```
containers: {
    db: {
        image: "docker.io/library/postgres:12.1"
        env: {
            "POSTGRES_USER": "postgres"
            "POSTGRES_PASSWORD": "postgres"
            "PGDATA": "/var/lib/postgresql/data/pgdata"
        }
        ports: publish: "5432/tcp"
        dirs: {
            "/var/lib/postgresql/data": "volume://pgdata"
        }
    }

    keycloak: {
        image: "europe-west4-docker.pkg.dev/qu-qu-353611/qu/qu-keycloak:dev"
        env: {
          "KEYCLOAK_ADMIN": "admin"
          "KEYCLOAK_ADMIN_PASSWORD": "admin"
          "KC_DB_URL_HOST": "db"
          "KC_DB_URL_DATABASE": "postgres"
          "KC_DB_USERNAME": "postgres"
          "KC_DB_PASSWORD": "postgres"
        }
        ports: publish: "8080/tcp"
        dependsOn: ["db"]
        cmd: ["start-dev", "--import-realm"]
    }
}

volumes: {
    "pgdata": {
        accessModes: "readWriteOnce"
     }
 }
```

I run using 
`acorn run --name qu -p mydb.com:db:5432 -p keycloak:8080 .`

but checking the status gives me 
```
acorn app
NAME      IMAGE          HEALTHY   UP-TO-DATE   CREATED     ENDPOINTS                                                                      MESSAGE
qu        3cf2dabfa57c   2         2            8m48s ago   tcp://<Pending Ingress> => db:5432, tcp://<Pending Ingress> => keycloak:8080   OK 
```

and ` minikube service list` gives this 
```
|------------------------------|------------------------------------|---------------|---------------------------|
|          NAMESPACE           |                NAME                |  TARGET PORT  |            URL            |
|------------------------------|------------------------------------|---------------|---------------------------|
| acorn-system                 | acorn-api                          | No node port  |
| acorn-system                 | registry                           | registry/5000 | http://192.168.49.2:32668 |
| awesome-acorn-c32b0fb1-19b   | db                                 | No node port  |
| awesome-acorn-c32b0fb1-19b   | keycloak                           | No node port  |
| crimson-paper-dfcaa0cd-13c   | db                                 | No node port  |
| crimson-paper-dfcaa0cd-13c   | keycloak                           | No node port  |
| default                      | kubernetes                         | No node port  |
| ingress-nginx                | ingress-nginx-controller           | http/80       | http://192.168.49.2:31147 |
|                              |                                    | https/443     | http://192.168.49.2:30974 |
| ingress-nginx                | ingress-nginx-controller-admission | No node port  |
| kube-system                  | kube-dns                           | No node port  |
| qu-9ad17f09-9b0              | db                                 | No node port  |
| qu-9ad17f09-9b0              | db-publish-9ad17f09-9b0            | 5432/5432     | http://192.168.49.2:30185 |
| qu-9ad17f09-9b0              | keycloak                           | No node port  |
| qu-9ad17f09-9b0              | keycloak-publish-9ad17f09-9b0      | 8080/8080     | http://192.168.49.2:31155 |
| solitary-breeze-3a6eba79-c0f | db                                 | No node port  |
| solitary-breeze-3a6eba79-c0f | keycloak                           | No node port  |
|------------------------------|------------------------------------|---------------|---------------------------|
```

I can actually access the DB on `http://192.168.49.2:30185`, but I want to access it on the port that was published !

Minikube has the ingress-plugin added already.

What Am I missing ?

**Comment by deathwaiting** [comment]
Ok, after some search, it seems this is a problem with minikube/kubernetes itself.
Setting a node port is usually just ignored and big-brother kubernetes decide a random port for you, and it seems there is no clear way to undo this.

Any way I ended up doing a simple python script that runs `minikube service list` and extract the generated url/IP's for the published services into a properties file, but that was more tailored to my project.

anyway it will be nice if this can be handled by acorn somehow, as too much time was wasted trying to understand all these kubernetes magic just to run a local dev environment.

**Comment by cjellick** [comment]
Right on. Thanks for opening the issue. There are definitely somethings we can do to improve this.

**Comment by cjellick** [comment]
*Publishing* a port makes it accessible outside of the cluster. Is that what you want to accomplish?

Seems to me that you might want the db port to only be accessible inside the cluster (by keycloak) and keycloak to be accessible as an http service. Assuming that's the case, you should do the following:
1. Change `ports: publish: "5432/tcp"` to just `ports: "5432/tcp"`
2. Remove `-p mydb.com:db:5432` from your run command entirely
3. Change `ports: publish: "8080/tcp"` to `ports: publish: "8080/http"`
4. I was going to tell you to also drop `-p keycloak:8080`, but I can't tell you what to do here definitively until I understand if you are trying to make keycloak accessible outside the cluster or only within the cluster

**Comment by deathwaiting** [comment]
> _Publishing_ a port makes it accessible outside of the cluster. Is that what you want to accomplish?
> 
> Seems to me that you might want the db port to only be accessible inside the cluster (by keycloak) and keycloak to be accessible as an http service. Assuming that's the case, you should do the following:
> 
> 1. Change `ports: publish: "5432/tcp"` to just `ports: "5432/tcp"`
> 2. Remove `-p mydb.com:db:5432` from your run command entirely
> 3. Change `ports: publish: "8080/tcp"` to `ports: publish: "8080/http"`
> 4. I was going to tell you to also drop `-p keycloak:8080`, but I can't tell you what to do here definitively until I understand if you are trying to make keycloak accessible outside the cluster or only within the cluster

Actually this script is for running a local development environment on my machine - and other dev's machines- .
In this case I need all the services to be published on local host or the cluster machine with fixed ports, so the service we are working on can access them, 
as it will run in debug mode for example on the developer machine directly not the cluster.

Having everything exposed is also needed so we can easily debug, we need to access the DB using DB tools to have better look.

**Comment by cjellick** [comment]
ok, makes sense.

We provide a little guidance here on how to get minikube setup: https://docs.acorn.io/installation/installing#local-development-clusters. Key takeaways are to install the ingress add-on and run `minikube tunnel`.

I am assuming your set up is minikube using the [docker driver](https://minikube.sigs.k8s.io/docs/drivers/). Correct me if I'm wrong.

After following those install instructions for minikube, I did a quick test of a modified version of your acornfile. I couldn't run your keycloak image, so here's the modified version I ran:
```
containers: {
    db: {
        image: "docker.io/library/postgres:12.1"
        env: {
            "POSTGRES_USER": "postgres"
            "POSTGRES_PASSWORD": "postgres"
            "PGDATA": "/var/lib/postgresql/data/pgdata"
        }
        ports: publish: "5432/tcp"
        dirs: {
            "/var/lib/postgresql/data": "volume://pgdata"
        }
    }
    app: {
        image: "nginx"
        ports: "80/http"
        scale: 1
        files: {
        // Simple index.html file
            "/usr/share/nginx/html/index.html": "<h1>test</h1>"
        }
    }
}

volumes: {
    "pgdata": {
        accessModes: "readWriteOnce"
     }
 }
```

Image build command:
```
acorn build -t test .
```

Run command:
```
acorn run -P test
```
That `-P` publishes all defined ports

Output of acorn ps:
```
patient-glitter   test      2         2            10m ago   http://app.patient-glitter.local.on-acorn.io => app:80, tcp://127.0.0.1:5432 => db:5432   OK
```

One limitation of this: only one app can consume port 5432, so if that port was already used by some other process or if you launched a second instance of the acorn, it would just hang in "pending" indefinitely.

After all that, I was able to connect to postgres:
```
$ psql --host 127.0.0.1 --port 5432 --user postgres
Password for user postgres:
psql (14.5, server 12.1 (Debian 12.1-1.pgdg100+1))
Type "help" for help.

postgres=#
```


One final note: my "app" service is on hostname rather than an IP because I defined it as "80/http" rather than "80/tcp". This gives us more flexibility in exposing it. If your keycloak service is indeed htttp, consider switching it from "8080/tcp" to "8080/http".

**Comment by cjellick** [comment]
FWIW, minikube, IMO, is not the easiest desktop k8s distribution to use. If you aren't married to it, try out Docker Desktop or Rancher Desktop. Docker Desktop does run into some licensing issues (not free once you're over a certain size, I think) so keep that in mind.

**Comment by cjellick** [comment]
Going to convert this to a discussion
